src/v/kafka/server/replicated_partition.cc

```diff
@@ -92,32 +92,34 @@ ss::future<model::record_batch_reader> replicated_partition::make_reader(
         ss::lw_shared_ptr<const storage::offset_translator_data> _translator;
     };
     auto rdr = co_await _partition->make_reader(cfg, deadline);
-    co_return model::make_record_batch_reader<reader>(
-      std::move(rdr).release(), _translator);
+    co_return storage::translating_reader(
+      model::make_record_batch_reader<reader>(
+        std::move(rdr).release(), _translator),
+      _translator);
 }
 
 ss::future<std::vector<cluster::rm_stm::tx_range>>
 replicated_partition::aborted_transactions(
-  model::offset base, model::offset last) {
-    model::offset local_kafka_start_offset = _translator->from_log_offset(
-      _partition->start_offset());
-    if (base < local_kafka_start_offset) {
-        // TODO: get offset translation information for the offsets range
-        // that we have read and use it to query
-        // _partition->aborted_transactions.
-        co_return std::vector<cluster::rm_stm::tx_range>{};
-    }
-
+  model::offset base,
+  model::offset last,
+  ss::lw_shared_ptr<const storage::offset_translator_data> ot_data) {
+    vassert(ot_data, "ntp {}: offset translator data must be present", ntp());
+
+    // Note: here we expect that local _partition contains aborted transaction
+    // ids for both local and remote offset ranges. This is true as long as
+    // rm_stm state has not been reset (for example when there is a partition
+    // transfer or when a stale replica recovers its log from beyond the log
+    // eviction point). See https://github.com/vectorizedio/redpanda/issues/3001
     auto source = co_await _partition->aborted_transactions(
-      _translator->to_log_offset(base), _translator->to_log_offset(last));
+      ot_data->to_log_offset(base), ot_data->to_log_offset(last));
```

Comment:
if I understand correctly, the `offset_translator_data` object learns about the gaps while we're scanning the segments. But what if we didn't scan the [base, last] range before? How it will handle the situation?

Reply:
That's the range that we just read (see `fetch.cc::read_from_partition`) so translation should be correct.